### DISCUSSION TITLE
Support creating msgpack request bodies

### DISCUSSION THREAD

**Discussion by andrei-dascalu** [Kong/insomnia#4505] [Ideas]
Hello,

I have a number of projects using MessagePack encoding as opposed to JSON. [MessagePack](https://msgpack.org/index.html) is practically a binary representation of JSON, so in a way it sits in the middle between the flexible/verbose/human-readable JSON and the strict/binary protocol buffers.

It would be great for my use cases to have support for creating MessagePack payloads to be send via POST/PUT (at least that's my main use case).

I could envision this happening in basically 2 ways:
- have a `msgpack` option in addition to JSON for the body content. The user would provide JSON content which Isomnia would encode appropriately before sending
- have a `msgpack` option which would allow the user to select a binary file whose contents would be used as request body contents.

Both options would result in the `content-type` header set to `application/x-msgpack`

For interpreting responses, it could behave in a similar way:

- when `content-type: application/x-msgpack` is received, the body would be interpreted as a binary payload and decoded appropriately.

Is this feasible? Does anyone have similar use cases?

Cheers!

**Comment by wdawson** [comment]
This feels like a great opportunity for an extension to Insomnia's plugin APIs.

**Comment by enghitalo** [comment]
Ping

**Comment by dinkopehar** [comment]
Hi @andrei-dascalu 

I found this discussion interesting. I faced similar problem as you, and while exploring, I didn't found any solution to this. So I created a plugin in the similar way you described above. A plugin can be found at https://github.com/dinko-pehar/insomnia-plugin-msgpack.

---

Both request and response are serialized/deserialized when certain criteria is matched. A plugin can be toggled **ON** or **OFF** . For example, if your endpoint requires MsgPack payload, toggling this plugin to **ON** and sending JSON request would automatically convert it to MsgPack format. In that way, highlight of JSON is preserved. Similar to that, response is automatically deserialized into JSON.
I tested it in few cases and it's working for me. You can install it and maybe give initial feedback does it work properly or do you need something to be fixed or expanded.
Thank you.
